[Added] Hashicorp Vault authentication options

dangerousplay opened the pull request:
Hello, I hope you are having a good day.

Implemented a variety of Vault authentication mechanisms such as **Kubernetes**, **AppRole**, and **JWT**. These can be utilized when an explicit token is not provided, offering flexibility to the users which don't need to log in before on Vault.

Have a nice day :)

dangerousplay committed:
feat(hashi_vault): authentication using Kubernetes, AppRole, JWT

Implemented authentication on Hashicorp Vault using Kubernetes, AppRole, JWT configurable by environment variables or novops yaml configuration file.

PierreBeucher commented:
That's awesome, thanks ! I'll review it ASAP

PierreBeucher commented:
It seems you're still actively working on this, I'll get on it once you pass PR in Ready state. Let me know if I can help :)

PierreBeucher commented on tests/test_hvault.rs:
```diff
@@ -186,7 +216,255 @@ async fn test_hashivault_client_address_load() -> Result<(), anyhow::Error> {
+        &jwt_key.to_encoding_key(),
+    )?;
+
+    env::set_var("VAULT_AUTH_JWT_TOKEN", token);
```
I can't make this test pass on my side, it seems this variable isn't used as `client.rs` will always look for file path locally. You probably wanted the client to use this variable somehow?

dangerousplay replied:
Yes, I fixed the code to use this constant if the environment variable doesn't exists.

PierreBeucher commented on src/modules/hashivault/client.rs:
```diff
@@ -1,13 +1,20 @@
+use crate::modules::hashivault::config::{HashiVaultAuth};
+
+
+const KUBERNETES_SA_JWT_PATH: &str = "/var/run/secrets/kubernetes.io/serviceaccount/token";
```
It seems this file is always used by default, ignoring the environment variable of the same name. Maybe you want to use env var here as well as default fle path?

dangerousplay replied:
Yes, I fixed the code to use this constant if the environment variable doesn't exists.

PierreBeucher commented on src/modules/hashivault/config.rs:
```diff
@@ -13,6 +13,26 @@ pub struct HashiVaultInput {
 
+#[derive(Debug, Deserialize, Clone, PartialEq, JsonSchema)]
+#[serde(tag = "type")]
+pub enum HashiVaultAuth {
```
Is it possible to have an `Option<String>` here with the "usual" path? eg. `auth/kubernetes` for Kubernetes, etc. 

Not blocking for merge, I can update it myself later :)

dangerousplay replied:
Yes, I changed to use default values for the mount path.

PierreBeucher commented:
LGTM for most things, thanks a lot for all these features ! A few things before merging:

- Can you rebase with [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) using `feat: xxx` and such? It will help generate release logs. I see you already did use similar convention, it should not be long :)
- Added a few comments for a failing test

dangerousplay commented:
Thank you for the code review.

I will increment the code to support to informing the Vault auth role, AppRole role ID, mount path by environment variables. I will add default values for mount path.

dangerousplay commented:
I hope you are having a good day ^^

I fixed the Vault authentication test that uses Kubernetes.
I changed the code to use default values for mount path depending on the authentication type. 
I rebased and squashed the commits using the conventional commit.

Sorry for the late update, unfortunately in the last week our city in Brazil was affected by one of the biggest water flood we have ever seen. Fortunately, we are secure and safe.

PierreBeucher commented:
Hey, thanks for the fixes ! I'm sorry for this difficult situation you had, hopefully you're better now... Please don't be sorry, you made an awesome contribution and there is nothing to be late for.

All tests are now passing on my end. I have a few more checks to run before merging but LGTM :) Thanks again

PierreBeucher commented:
Finally merged, sorry for the delay these last weeks I hadn't had much time for Novops. I'm gonna do a release soon with these new features, thanks again !
